irc-proto/src/prefix.rs

```diff
@@ -0,0 +1,204 @@
+//! A module providing an enum for a message prefix.
+use std::borrow::ToOwned;
+use std::string;
+use std::str::FromStr;
+use std::fmt;
+
+/// The Prefix indicates "the true origin of the message", according to the server.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Prefix {
+    /// servername, e.g. collins.mozilla.org
+    ServerName(String),
+    /// nickname [ ["!" username] "@" hostname ]
+    /// i.e. Nickname(nickname, username, hostname)
+    /// Any of the strings may be ""
+    Nickname(String, String, String),
+}
+
+impl Prefix {
+    /// Creates a prefix by parsing a string.
+    ///
+    /// # Example
+    /// ```
+    /// # extern crate irc;
+    /// # use irc::client::prelude::*;
+    /// # fn main() {
+    /// Prefix::new_from_str("nickname!username@hostname");
+    /// Prefix::new_from_str("example.com");
+    /// # }
+    /// ```
+    pub fn new_from_str(s: &str) -> Prefix {
+        #[derive(Copy, Clone, Eq, PartialEq)]
+        enum Active {
+            Name = 0,
+            User = 1,
+            Host = 2,
+        }
+
+        let mut name = String::new();
+        let mut user = String::new();
+        let mut host = String::new();
+        let mut active = Active::Name;
+
+        for c in s.chars() {
+            match c {
+                // We consider the '.' to be a ServerName except if a ! has already
+                // been encountered.
+                '.' if active == Active::Name => return Prefix::ServerName(s.to_owned()),
```

Comment:
This would result in `a.b.c!foo@bar` being considered a server name, right? According to the grammar, server names can't include `!` or `@`.

(According to the grammar, nicknames can't include `.`, so `a.b.c!foo@bar` wouldn't be a valid prefix at all, but IRC daemons tend to be much more lenient about nicknames than the grammar is, so I wouldn't be surprised if there are nicknames out there with `.` in them.)